truffle/src/main/java/org/jruby/truffle/runtime/rope/RopeOperations.java

```diff
@@ -144,4 +131,216 @@ public static int strLength(Encoding enc, byte[] bytes, int p, int end) {
         return StringSupport.strLength(enc, bytes, p, end);
     }
 
+    public static LeafRope flatten(Rope rope) {
+        if (rope instanceof LeafRope) {
+            return (LeafRope) rope;
+        }
+
+        return create(flattenBytes(rope), rope.getEncoding(), rope.getCodeRange());
+    }
+
+    @TruffleBoundary
+    /**
+     * Performs an iterative depth first search of the Rope tree to calculate its byte[] without needing to populate
+     * the byte[] for each level beneath. Every LeafRope has its byte[] populated by definition. The goal is to determine
+     * which descendant LeafRopes contribute bytes to the top-most Rope's logical byte[] and how many bytes they should
+     * contribute. Then each such LeafRope copies the appropriate range of bytes to a shared byte[].
+     *
+     * Rope trees can be very deep. An iterative algorithm is preferable to recursion because it removes the overhead
+     * of stack frame management. Additionally, a recursive algorithm will eventually overflow the stack if the Rope
+     * tree is too deep.
+     */
+    public static byte[] flattenBytes(Rope rope) {
```

Comment:
This seems quite complex. So in essence it's a DFS walk on the rope?
Can't we just call some SubstringRope.copy(buffer, startIndex)? That could recurse a bit but anyway the depth is bounded, isn't it?